modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/nhttp/NhttpSharedOutputBuffer.java

```diff
@@ -0,0 +1,375 @@
+                pos += chunk;
+            }
+        } finally {
+            this.lock.unlock();
+        }
+    }
+
+    public void write(final byte[] b) throws IOException {
+        if (b == null) {
+            return;
+        }
+        write(b, 0, b.length);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public void write(final int b) throws IOException {
+        this.lock.lock();
+        try {
+            Asserts.check(!this.shutdown && !this.endOfStream, "Buffer already closed for writing");
+            // Set buffer to write mode
+            setInputMode();
+            // At the moment buffer is in the write mode. So position equals to capacity.
+            // if position >= limit(= capacity) -> true i.e. buffer is full
+            // If this buffer is full, that means the data should be written to output buffer (Need to flush).
+            if (!this.buffer.hasRemaining()) {
+                flushContent();
+                // If "awaitInterrupted" was set to false, that means something went wrong(timeout happened)
+                // therefore, we clear the buffer and stop the flow with an exception.
+                if (!awaitInterrupted) {
+                    this.buffer.clear();
+                    throw new IOException("Output buffer write time out exceeded");
+                }
+                // Set buffer to write mode. (Inside flushContent method, buffer was set to read mode)
+                setInputMode();
+            }
+            this.buffer.put((byte)b);
+        } finally {
+            this.lock.unlock();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public void flush() throws IOException {
+    }
+
+    /**
+     * Flushes the content in the buffer if any.
+     *
+     * @throws IOException Thrown if the thread was interrupted while flushing
+     */
+    private void flushContent() throws IOException {
+        this.lock.lock();
+        try {
+            try {
+                // hasData sets the buffer to read mode and calls hasRemaining method and returns the result.
+                // if position (= initially 0, incremented as data is read) < limit(= how much bytes have been written
+                // to the buffer) -> true i.e. there is data to be read
+                while (super.hasData()) {
+                    if (this.shutdown) {
+                        throw new InterruptedIOException("Output operation aborted");
+                    }
+                    // Request event notifications to be triggered when the underlying
+                    // channel is ready for output operations.
+                    if (this.ioctrl != null) {
+                        this.ioctrl.requestOutput();
+                    }
+                    // Ask the thread which is reading this buffer to wait for a notification to
+                    // write data out
+                    awaitInterrupted = this.condition.await(timeout, TimeUnit.MILLISECONDS);
```

Comment:
what happens to reading threads if awaitInterrupted=true? Do they get notified?

Reply:
awaitInerrupted becomes true when the thread is notified.